### PR TITLE
Log4 j2 913 - Add support for HTTP/S configuration monitoring. Support Spring Cloud Config

### PR DESCRIPTION
**Pull request by rgoers**
For your initial review. I still need to update the web site and look into spring cloud push notifications, so don't merge this yet.

### CODE REVIEW

File: log4j-core/src/main/java/org/apache/logging/log4j/core/config/ConfigurationFactory.java [hunk 1]
```diff
@@ -294,7 +302,22 @@ static String extractClassLoaderUriPath(final URI uri) {
...
         try {
             final URL url = new URL(config);
-            return new ConfigurationSource(url.openStream(), FileUtils.fileFromUri(url.toURI()));
+            URLConnection urlConnection = url.openConnection();
+            if (url.getProtocol().equals(HTTPS)) {
+                SslConfiguration sslConfiguration = SslConfigurationFactory.getSslConfiguration();
+                if (sslConfiguration != null) {
+                    ((HttpsURLConnection) urlConnection).setSSLSocketFactory(sslConfiguration.getSslSocketFactory());
+                    if (!sslConfiguration.isVerifyHostName()) {
+                        ((HttpsURLConnection) urlConnection).setHostnameVerifier(LaxHostnameVerifier.INSTANCE);
```
Comment (jvz) [inline log4j-core/src/main/java/org/apache/logging/log4j/core/config/ConfigurationFactory.java]: I mean, at least you're not setting it globally, but why is this here?

File: log4j-core/src/main/java/org/apache/logging/log4j/core/config/ConfigurationSource.java [hunk 2]
```diff
@@ -79,6 +98,24 @@ public ConfigurationSource(final InputStream stream, final URL url) {
...
+
+    /**
+     * Constructs a new {@code ConfigurationSource} with the specified input stream that originated from the specified
+     * url.
+     *
+     * @param stream the input stream
+     * @param url the URL where the input stream originated
+     * @param lastModified when the source was last modified.
+     */
+    public ConfigurationSource(final InputStream stream, final URL url, long lastModified) {
```
Comment (jvz) [inline log4j-core/src/main/java/org/apache/logging/log4j/core/config/ConfigurationSource.java]: So I'll note here that for 3.x, we should certainly be redoing this API. It exposes way too much internal state.

File: log4j-core/src/main/java/org/apache/logging/log4j/core/config/HttpWatcher.java [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+    private static final String XML = "application/xml";
+    private static final String PROPERTIES = "text/x-java-properties";
+    private static final String TEXT = "text/plain";
+    private static final int NOT_MODIFIED = 304;
+    private static final int OK = 200;
+    private static final int BUF_SIZE = 1024;
+    private static final String HTTP = "http";
+    private static final String HTTPS = "https";
+
+    public HttpWatcher(final Configuration configuration, final Reconfigurable reconfigurable,
```
Comment (jvz) [inline log4j-core/src/main/java/org/apache/logging/log4j/core/config/HttpWatcher.java]: Should this be private?

File: log4j-core/src/main/java/org/apache/logging/log4j/core/net/ssl/SslConfigurationFactory.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+/**
+ * Creates an SSL configuration from Log4j properties.
+ */
+public class SslConfigurationFactory {
+
+    private static final Logger LOGGER = StatusLogger.getLogger();
+    private static SslConfiguration sslConfiguration = null;
+
+    private static final String trustStorelocation = "log4j2.trustStore.location";
+    private static final String trustStorePassword = "log4j2.trustStore.password";
```
Comment (jvz) [inline log4j-core/src/main/java/org/apache/logging/log4j/core/net/ssl/SslConfigurationFactory.java]: While this will resolve to `log4j2.trustStorePassword` at runtime, you should at least stay consistent in your naming scheme within this single class (see `log4j2.keyStorePassword`).

File: log4j-core/src/main/java/org/apache/logging/log4j/core/util/Watcher.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+import org.apache.logging.log4j.core.config.Reconfigurable;
+
+/**
+ * Watches for changes in a Source and performs an action when it is modified.
+ *
+ * @see WatchManager
+ */
+public interface Watcher {
+
+    public static final String CATEGORY = "Watcher";
```
Comment (jvz) [inline log4j-core/src/main/java/org/apache/logging/log4j/core/util/Watcher.java]:
```suggestion
    String CATEGORY = "Watcher";
```

The three keywords are all implied by a "field" in an interface.

File: log4j-core/src/main/java/org/apache/logging/log4j/core/util/WatcherFactory.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+            return null;
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    public static <T extends Watcher> T instantiate(String name, final Class<T> clazz, final Configuration configuration,
+        final Reconfigurable reconfigurable, final List<ConfigurationListener> listeners, long lastModifiedMillis) {
+        Objects.requireNonNull(clazz, "No class provided");
+        try {
+            Constructor constructor = clazz.getConstructor(Configuration.class, Reconfigurable.class, List.class, long.class);
```
Comment (jvz) [inline log4j-core/src/main/java/org/apache/logging/log4j/core/util/WatcherFactory.java]:
```suggestion
            Constructor<T> constructor = clazz.getConstructor(Configuration.class, Reconfigurable.class, List.class, long.class);
```

That way you don't need a generic unchecked cast below.

File: log4j-core/src/main/java/org/apache/logging/log4j/core/util/WrappedFileWatcher.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) 2019 Nextiva, Inc. to Present.
```
Comment (jvz) [inline log4j-core/src/main/java/org/apache/logging/log4j/core/util/WrappedFileWatcher.java]: Needs updating.

File: log4j-core/src/main/java/org/apache/logging/log4j/core/util/WatcherFactory.java [hunk 8]
```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (c) 2019 Nextiva, Inc. to Present.
```
Comment (jvz) [inline log4j-core/src/main/java/org/apache/logging/log4j/core/util/WatcherFactory.java]: Needs updating.

File: log4j-core/src/test/java/org/apache/logging/log4j/core/util/WatchHttpTest.java [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+    private static final String XML = "application/xml";
+
+    private static final boolean IS_WINDOWS = PropertiesUtil.getProperties().isOsWindows();
+
+    @BeforeClass
+    public static void beforeClass() {
+        try {
+            formatter = FastDateFormat.getInstance("EEE, dd MMM yyyy HH:mm:ss", TimeZone.getTimeZone("UTC"));
+        } catch (Exception ex) {
+            System.err.println("Unable to create date format.");
```
Comment (jvz) [inline log4j-core/src/test/java/org/apache/logging/log4j/core/util/WatchHttpTest.java]: You could use `LowLevelLogUtil` here if you want.

File: log4j-core/src/test/java/org/apache/logging/log4j/core/util/WatchHttpTest.java [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+        final ConfigurationScheduler scheduler = new ConfigurationScheduler();
+        scheduler.incrementScheduledItems();
+        final WatchManager watchManager = new WatchManager(scheduler);
+        watchManager.setIntervalSeconds(1);
+        scheduler.start();
+        watchManager.start();
+        try {
+            watchManager.watch(new Source(url.toURI(), previous.getTimeInMillis()), new HttpWatcher(configuration, null,
+                listeners, previous.getTimeInMillis()));
+            Thread.sleep(1500);
```
Comment (jvz) [inline log4j-core/src/test/java/org/apache/logging/log4j/core/util/WatchHttpTest.java]: Is there any way you could instead expose a `CoundDownLatch` for testing purposes? Then you could `await()` it instead of allowing any problems under CI load.

File: log4j-spring-cloud-config/log4j-spring-cloud-config-client/src/main/java/org/apache/logging/log4j/spring/cloud/config/client/Log4j2CloudConfigLoggingSystem.java [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+
+    @Override
+    protected String[] getStandardConfigLocations() {
+        String[] locations = super.getStandardConfigLocations();
+        PropertiesUtil props = new PropertiesUtil(new Properties());
+        String location = props.getStringProperty(ConfigurationFactory.CONFIGURATION_FILE_PROPERTY);
+        if (location != null) {
+            List<String> list = Arrays.asList(super.getStandardConfigLocations());
+            list.add(location);
+            locations = list.toArray(new String[list.size()]);
```
Comment (jvz) [inline log4j-spring-cloud-config/log4j-spring-cloud-config-client/src/main/java/org/apache/logging/log4j/spring/cloud/config/client/Log4j2CloudConfigLoggingSystem.java]: From the latest benchmarks I remember reading about, it's generally faster to just pass 0 here instead of the real size since `toArray()` has to check `list.size()` first anyways.

File: log4j-spring-cloud-config/log4j-spring-cloud-config-server/src/main/java/org/apache/logging/log4j/spring/cloud/config/server/controller/Log4jResourceController.java [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+ * @author Dave Syer
+ * @author Daniel Lavoie
+ *
+ */
+@RestController
+@RequestMapping(method = RequestMethod.GET, path = "${spring.cloud.config.server.prefix:}/resource")
+public class Log4jResourceController {
+
+    @Autowired
+	private ResourceRepository resourceRepository;
```
Comment (jvz) [inline log4j-spring-cloud-config/log4j-spring-cloud-config-server/src/main/java/org/apache/logging/log4j/spring/cloud/config/server/controller/Log4jResourceController.java]: Seems like this file has a mix of tabs and spaces.

File: log4j-spring-cloud-config/log4j-spring-cloud-config-server/src/main/java/org/apache/logging/log4j/spring/cloud/config/server/controller/Log4jResourceController.java [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+        name = resolveName(name);
+        label = resolveLabel(label);
+        Resource resource = this.resourceRepository.findOne(name, profile, label, path);
+        if (checkNotModified(request, resource)) {
+            // Content was not modified. Just return.
+            return null;
+        }
+        // ensure InputStream will be closed to prevent file locks on Windows
+        try (InputStream is = resource.getInputStream()) {
+            String text = StreamUtils.copyToString(is, Charset.forName("UTF-8"));
```
Comment (jvz) [inline log4j-spring-cloud-config/log4j-spring-cloud-config-server/src/main/java/org/apache/logging/log4j/spring/cloud/config/server/controller/Log4jResourceController.java]: Can use `StandardCharsets.UTF_8` here.

File: log4j-core/src/main/java/org/apache/logging/log4j/core/config/ConfigurationFactory.java (hunk 1 above)
Comment (rgoers) [reply log4j-core/src/main/java/org/apache/logging/log4j/core/config/ConfigurationFactory.java]: HostNameVerifier? I copied this code from the HttpAppender but since we haven't configured yet I had to get creative in how the SSL settings would be obtained.

File: log4j-core/src/main/java/org/apache/logging/log4j/core/config/ConfigurationSource.java (hunk 2 above)
Comment (rgoers) [reply log4j-core/src/main/java/org/apache/logging/log4j/core/config/ConfigurationSource.java]: Good luck with that. All this information has to be passed from one configuration to another to allow reconfiguration to occur.

File: log4j-core/src/main/java/org/apache/logging/log4j/core/config/HttpWatcher.java (hunk 3 above)
Comment (rgoers) [reply log4j-core/src/main/java/org/apache/logging/log4j/core/config/HttpWatcher.java]: No. It is called by unit tests.

File: log4j-core/src/main/java/org/apache/logging/log4j/core/net/ssl/SslConfigurationFactory.java (hunk 4 above)
Comment (rgoers) [reply log4j-core/src/main/java/org/apache/logging/log4j/core/net/ssl/SslConfigurationFactory.java]: Fixed

File: log4j-core/src/main/java/org/apache/logging/log4j/core/util/Watcher.java (hunk 5 above)
Comment (rgoers) [reply log4j-core/src/main/java/org/apache/logging/log4j/core/util/Watcher.java]: Fixed

File: log4j-core/src/test/java/org/apache/logging/log4j/core/util/WatchHttpTest.java (hunk 9 above)
Comment (rgoers) [reply log4j-core/src/test/java/org/apache/logging/log4j/core/util/WatchHttpTest.java]: It is just a unit test. I am not sure what using LowLevelLogUtil would gain me here.

File: log4j-core/src/test/java/org/apache/logging/log4j/core/util/WatchHttpTest.java (hunk 10 above)
Comment (rgoers) [reply log4j-core/src/test/java/org/apache/logging/log4j/core/util/WatchHttpTest.java]: Turns out I didn't need to sleep at all anyway...

File: log4j-spring-cloud-config/log4j-spring-cloud-config-server/src/main/java/org/apache/logging/log4j/spring/cloud/config/server/controller/Log4jResourceController.java (hunk 12 above)
Comment (rgoers) [reply log4j-spring-cloud-config/log4j-spring-cloud-config-server/src/main/java/org/apache/logging/log4j/spring/cloud/config/server/controller/Log4jResourceController.java]: This file was acquired from Spring. They use tabs. I'll reformat it.

File: log4j-spring-cloud-config/log4j-spring-cloud-config-client/src/main/java/org/apache/logging/log4j/spring/cloud/config/client/Log4j2CloudConfigLoggingSystem.java (hunk 11 above)
Comment (rgoers) [reply log4j-spring-cloud-config/log4j-spring-cloud-config-client/src/main/java/org/apache/logging/log4j/spring/cloud/config/client/Log4j2CloudConfigLoggingSystem.java]: I had always thought that if you pass in an array that toArray would just use it.

File: log4j-spring-cloud-config/log4j-spring-cloud-config-server/src/main/java/org/apache/logging/log4j/spring/cloud/config/server/controller/Log4jResourceController.java (hunk 13 above)
Comment (rgoers) [reply log4j-spring-cloud-config/log4j-spring-cloud-config-server/src/main/java/org/apache/logging/log4j/spring/cloud/config/server/controller/Log4jResourceController.java]: Fixed

File: log4j-core/src/main/java/org/apache/logging/log4j/core/config/ConfigurationFactory.java (hunk 1 above)
Comment (rgoers) [reply log4j-core/src/main/java/org/apache/logging/log4j/core/config/ConfigurationFactory.java]: It is handled in SslConfigurationFactory.

File: log4j-spring-cloud-config/log4j-spring-cloud-config-client/src/main/java/org/apache/logging/log4j/spring/cloud/config/client/Log4j2CloudConfigLoggingSystem.java (hunk 11 above)
Comment (jvz) [reply log4j-spring-cloud-config/log4j-spring-cloud-config-client/src/main/java/org/apache/logging/log4j/spring/cloud/config/client/Log4j2CloudConfigLoggingSystem.java]: The key here is that even if you pass it an array, toArray has to check the length of the passed array against the length of the array being copied. toArray will resize the array if needed, so there's less if-checks executed in the array of length 0 execution path. Basically, toArray() is programmed to construct the new array regardless, so save yourself the bytecode.

Comment (rgoers) [reply log4j-spring-cloud-config/log4j-spring-cloud-config-client/src/main/java/org/apache/logging/log4j/spring/cloud/config/client/Log4j2CloudConfigLoggingSystem.java]: You must be thinking of a different method. The javadoc for this says "Returns a fixed-size list backed by the specified array. (Changes to the returned list "write through" to the array.)". Looking at the implementation that is exactly what it does. Passing an array of size 0 would require me to manually populate the list from the array, which is the whole point of this method.

Comment (jvz) [reply log4j-spring-cloud-config/log4j-spring-cloud-config-client/src/main/java/org/apache/logging/log4j/spring/cloud/config/client/Log4j2CloudConfigLoggingSystem.java]: Oh I think you're right. Found more info [here](https://shipilev.net/blog/2016/arrays-wisdom-ancients/) and it doesn't seem to be an issue in this scenario (or in general anymore; this old issue seems to have been fixed). The key difference here is that calling `size()` on an array list is `O(1)`, so it's not as big an issue. If you were calling, say, `Collection::toArray` on a collection that doesn't cache its size (random glance at some standard collection classes indicates they normally do cache that), or worse, if the collection changed between checking the size and calling toArray, then that causes the performance latency.

File: log4j-core/src/main/java/org/apache/logging/log4j/core/appender/rolling/CompositeTriggeringPolicy.java [hunk 14]
```diff
@@ -50,6 +50,7 @@ private CompositeTriggeringPolicy(final TriggeringPolicy... triggeringPolicies)
     @Override
     public void initialize(final RollingFileManager manager) {
         for (final TriggeringPolicy triggeringPolicy : triggeringPolicies) {
+            LOGGER.debug("Initializing triggering policy ", triggeringPolicy.toString());
```
Comment (cakofony) [inline log4j-core/src/main/java/org/apache/logging/log4j/core/appender/rolling/CompositeTriggeringPolicy.java]: nit: I don't think toString isn't necessary here, the logger will get the string value if debug is enabled

File: log4j-core/src/main/java/org/apache/logging/log4j/core/config/ConfigurationSource.java (hunk 2 above)
Comment (jvz) [reply log4j-core/src/main/java/org/apache/logging/log4j/core/config/ConfigurationSource.java]: I've considered in the past porting over or implementing a subset of the [`Resource`](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/core/io/Resource.html) interface from Spring. Do you think such an API would be more useful than having to inspect the ConfigurationSource union-like class?
